src/main/kotlin/com/dp/advancedgunnerycontrol/weaponais/AGCUitls.kt

```diff
@@ -243,3 +244,35 @@ fun mapBooleanToSpecificString(boolValue: Boolean, trueString: String, falseStri
 internal infix fun Vector2f.times_(d: Float): Vector2f {
     return Vector2f(d * x, d * y)
 }
+
+
+/**
+ * @return approximate angular distance of target from current weapon facing in rad
+ * note: approximation works well for small values and is off by a factor of PI/2 for 180°
+ * @param entity: Relative coordinates (velocity-compensated)
+ */
+fun angularDistanceFromWeapon(entity: Vector2f, weapon: WeaponAPI): Float {
+    val weaponDirection = vectorFromAngleDeg(weapon.currAngle)
+    val distance = entity - weapon.location
+    val entityDirection = distance times_ (1f / distance.length())
+    return (weaponDirection - entityDirection).length()
+}
+
+fun linearDistanceFromWeapon(entity: Vector2f, weapon: WeaponAPI): Float {
+    return (weapon.location - entity).length()
+}
+/**
+ * @param entity: In relative coordinates
+ * @param collRadius: Include any tolerances in here
+ * @param aimPoint: Point the weapon is aiming at, deduced from current weapon facing if not provided
+ */
+fun determineIfShotWillHit(entity: Vector2f, collRadius: Float, weapon: WeaponAPI, aimPoint: Vector2f? = null) : Boolean{
```

Comment:
Ok, this took me a while to wrap my head around xD
If I understand this correctly, x is essentially the cosine times the length of e divided by the length of p.

So by multiplying p with x, you first scale it to the length of e (by dividing it by its own length and multiplying it with the length of e).  That makes sense, p has arbitrary length (often 1).

Then you scale its length with the cosine. The scaling with the cosine is, colloquially put, to get the portion of the projectile trajectory before it passes the center of the target entity (or, in other words, the point closest to the target entity on the projectile trajectory).

In the return statement you first make sure that the weapon isn't pointing in the opposite direction (you could also take the absolute value of x instead of performing the x > 0 check, correct? The x > 0 is simply faster) and then check that the distance squared between the closest point to the target entity and the center of the target entity is less than its collision radius.

Is that interpretation of this function correct?

If possible, it would be awesome to have a comment or two explaining these formulas =) (I know, the code base is full of uncommented computations that could use a comment or two xD)

Reply:
Now that you mention it, `x` does indeed look like the dot product of `e` and `p` divided by `lengthSquared(p)`. But I don't know why the solution has this form and how to interpret it geometrically.

I interpret `x` as the minimum point of function `f(x)=distSquared(aimPoint * x, entity)`. This function describes the distance between the target and the projectile along the projectile path. We know `f(x)` is a positive quadratic function, so it has one minimum, which can be found by looking the correct formula on wikipedia :D

And indeed, `aimPoint` multiplied by the minimum `x` is the point where the projectile passes the target and is closest to it. So the rest of your interpretation is correct.

I'll add some comments, and possibly rename `x` to avoid confusion with x argument in `f(x)` formula.

Reply:
Awesome, thanks! 

I might be wrong, but I think my "geometrical" explanation is actually almost the same, just using the "a straight orthogonal line is the shortest path" approach. If you imagine a triangle where one point is at the weapon, one point at the center of the target and the third point is placed so that it's a 90° angle, that point is the closest point to the center of the target entity. And thus the distance to that point from the weapon is cos(alpha) * distance_to_target. So by scaling p to the length of e and multiplying by the cos(alpha), you get the closest point. Anyhow, thank you for the explanation, that definitely makes sense and is easier to follow than the geometric interpretation xD